google-ads-examples/src/main/java/com/google/ads/googleads/examples/advancedoperations/AddDynamicSearchAds.java

```diff
@@ -0,0 +1,335 @@
+      }
+    }
+  }
+
+  /**
+   * Runs the example.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @throws GoogleAdsException if an API request failed with one or more service errors.
+   */
+  private static void runExample(GoogleAdsClient googleAdsClient, long customerId) {
+    String budgetResourceName = addCampaignBudget(googleAdsClient, customerId);
+    String campaignResourceName = addCampaign(googleAdsClient, customerId, budgetResourceName);
+    String adGroupResourceName = addAdGroup(googleAdsClient, customerId, campaignResourceName);
+    addExpandedDSA(googleAdsClient, customerId, adGroupResourceName);
+    addWebPageCriteria(googleAdsClient, customerId, adGroupResourceName);
+  }
+
+  /**
+   * Adds a campaign budget.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @return the campaign budget resource name.
+   */
+  private static String addCampaignBudget(GoogleAdsClient googleAdsClient, long customerId) {
+    // Creates the budget.
+    CampaignBudget campaignBudget =
+        CampaignBudget.newBuilder()
+            .setName(StringValue.of("Interplanetary Cruise Budget #" + System.currentTimeMillis()))
+            .setAmountMicros(Int64Value.of(3_000_000))
+            .setDeliveryMethod(BudgetDeliveryMethod.STANDARD)
+            .build();
+
+    // Creates the operation.
+    CampaignBudgetOperation operation =
+        CampaignBudgetOperation.newBuilder().setCreate(campaignBudget).build();
+
+    // Creates the campaign budget service client.
+    try (CampaignBudgetServiceClient campaignBudgetServiceClient =
+        googleAdsClient.getLatestVersion().createCampaignBudgetServiceClient()) {
+      // Adds the campaign budget.
+      MutateCampaignBudgetsResponse response =
+          campaignBudgetServiceClient.mutateCampaignBudgets(
+              Long.toString(customerId), ImmutableList.of(operation));
+      // Displays the result.
+      String budgetResourceName = response.getResults(0).getResourceName();
+      System.out.printf("Added budget with resource name '%s'.%n", budgetResourceName);
+      return budgetResourceName;
+    }
+  }
+
+  /**
+   * Adds a campaign.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param budgetResourceName the campaign budget resource name.
+   * @return the campaign resource name.
+   */
+  private static String addCampaign(
+      GoogleAdsClient googleAdsClient, long customerId, String budgetResourceName) {
+    // Creates the campaign.
+    Campaign campaign =
+        Campaign.newBuilder()
+            .setName(StringValue.of("Interplanetary Cruise #" + System.currentTimeMillis()))
+            .setAdvertisingChannelType(AdvertisingChannelType.SEARCH)
+            .setStatus(CampaignStatus.PAUSED)
+            .setManualCpc(ManualCpc.newBuilder().build())
```

Comment:
Do you need to set the bid on the ManualCpc object?

Reply:
I don't believe so because this line just sets the bidding strategy type, not the actual CPC (see [campaign](https://developers.google.com/google-ads/api/reference/rpc/google.ads.googleads.v1.resources#campaign) and [manual CPC](https://developers.google.com/google-ads/api/reference/rpc/google.ads.googleads.v1.common#google.ads.googleads.v1.common.ManualCpc))

Reply:
Sorry, I was thinking of other strategy types.